hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/replicas/ReplicaStateVerifier.java

```diff
@@ -0,0 +1,221 @@
+  private final ContainerOperationClient containerOperationClient;
+  private final XceiverClientManager xceiverClientManager;
+  // cache for replica details from the DNs
+  private final Cache<ReplicaKey, ContainerDataProto> containerDataCache = CacheBuilder.newBuilder()
+      .maximumSize(1000)
+      .build();
+  // cache for container info and encodedToken from the SCM
+  private final Cache<Long, ContainerInfoToken> encodedTokenCache = CacheBuilder.newBuilder()
+      .maximumSize(1000)
+      .build();
+
+  public ReplicaStateVerifier(OzoneConfiguration conf) throws IOException {
+    containerOperationClient = new ContainerOperationClient(conf);
+    xceiverClientManager = containerOperationClient.getXceiverClientManager();
+  }
+
+  @Override
+  public String getType() {
+    return CHECK_TYPE;
+  }
+
+  @Override
+  public BlockVerificationResult verifyBlock(DatanodeDetails datanode, OmKeyLocationInfo keyLocation,
+                                             int replicaIndex) {
+    try {
+      StringBuilder replicaCheckMsg = new StringBuilder().append("Replica state is ");
+      boolean pass = false;
+
+      ContainerInfoToken containerInfoToken = getContainerInfoToken(keyLocation.getContainerID());
+      ContainerInfo containerInfo = containerInfoToken.getContainerInfo();
+
+      ContainerDataProto containerData = getContainerData(datanode, keyLocation, replicaIndex, containerInfoToken);
+      if (containerData.getState().equals(ContainerDataProto.State.UNHEALTHY)) {
+        replicaCheckMsg.append("UNHEALTHY");
+      } else if (containerData.getState().equals(ContainerDataProto.State.INVALID)) {
+        replicaCheckMsg.append("INVALID");
+      } else if (containerData.getState().equals(ContainerDataProto.State.DELETED)) {
+        replicaCheckMsg.append("DELETED");
+      } else {
+        pass = true;
+      }
+      replicaCheckMsg.append(", Container state in SCM is ").append(containerInfo.getState());
+
+      if (pass) {
+        return BlockVerificationResult.pass();
+      } else {
+        return BlockVerificationResult.failCheck(replicaCheckMsg.toString());
+      }
+    } catch (IOException e) {
+      return BlockVerificationResult.failIncomplete(e.getMessage());
+    }
+  }
+
+  public ContainerDataProto getContainerData(DatanodeDetails dn, OmKeyLocationInfo keyLocation,
+                                             int replicaIndex, ContainerInfoToken containerInfoToken)
+      throws IOException {
+    long containerId = containerInfoToken.getContainerInfo().getContainerID();
+    ReplicaKey key = new ReplicaKey(dn, containerId);
+    ContainerDataProto cachedData = containerDataCache.getIfPresent(key);
+    if (cachedData != null) {
+      return cachedData;
+    }
+    // Cache miss - fetch and store
+    ContainerDataProto data = fetchContainerDataFromDatanode(dn, containerId, keyLocation,
+        replicaIndex, containerInfoToken);
+    containerDataCache.put(key, data);
```

Comment:
```suggestion
    if (data != null) {
  containerDataCache.put(key, data);
    }
```
If data is null by chance, we'll get a null value from the cache every time!